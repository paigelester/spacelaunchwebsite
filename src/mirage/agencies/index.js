import AgencySeed from './seed';
import AgencyModels from './models';
import AgencyRoutes from './routes';
import AgencyFactories from './factories';

export default {
    seed: AgencySeed,
    models: AgencyModels,
    routes: AgencyRoutes,
    factories: AgencyFactories
}