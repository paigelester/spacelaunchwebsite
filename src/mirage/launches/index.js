import LaunchSeed from './seed';
import LaunchModels from './models';
import LaunchRoutes from './routes';
import LaunchFactories from './factories';

export default {
    seed: LaunchSeed,
    models: LaunchModels,
    routes: LaunchRoutes,
    factories: LaunchFactories
}