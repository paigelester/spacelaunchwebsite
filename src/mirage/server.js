import { createServer, Model, Factory, trait, Response } from 'miragejs';

export const makeServer = () => {
    let server = createServer({

        models: {
            launch: Model
        },

        factories: {
            launch: Factory.extend({
                name(i) { return `Launch name ${i + 1}`; },
                image: 'https://spacelaunchnow-prod-east.nyc3.digitaloceanspaces.com/media/launch_images/starship2520sn_image_20210107083733.jpeg',
                status: {
                    name(i) {
                        let statuses = ['Go for launch', 'To be confirmed'];

                        return statuses[i % statuses.length];
                    },
                },
                window_start(i) {
                    const startDate = new Date();
                    startDate.setDate(startDate.getDate() + (i + 1));

                    return startDate;
                },
                window_end(i) {
                    const endDate = new Date();
                    endDate.setDate(endDate.getDate() + (i + 2));

                    return endDate;
                },
                launch_service_provider: {
                    name: 'SpaceX',
                    type: 'Commercial',
                    url: 'https://ll.thespacedevs.com/2.1.0/agencies/121/'
                },

                previous: trait({
                    window_start(i) {
                        const startDate = new Date();
                        startDate.setDate(startDate.getDate() - (i + 2));

                        return startDate;
                    },
                    window_end(i) {
                        const endDate = new Date();
                        endDate.setDate(endDate.getDate() - (i + 1));

                        return endDate;
                    }
                })
            })
        },

        seeds(dataServer) {
            dataServer.createList('launch', 10);
            dataServer.createList('launch', 10, 'previous');
        },

        routes() {
            this.urlPrefix = 'https://ll.thespacedevs.com/2.1.0/'

            this.get('launch/:id', (schema, request) => {
                const launchId = request.params.id;

                const launch = schema.launches.find(launchId);
                
                if (launch === null) {
                    return Response(404);
                }

                return launch.attrs;
            });

            this.get("launch/upcoming", (schema, request) => {
                const limit = request.queryParams.limit;

                const upcomingLaunches = schema.launches.all().models.reduce((prev, current) => {
                    if (current.window_start > new Date()) {
                        prev.push(current);
                    }

                    return prev;
                }, []).slice(0, limit);

                return {
                    count: upcomingLaunches.length,
                    results: upcomingLaunches
                };
            });
        }

    });

    return server;
};