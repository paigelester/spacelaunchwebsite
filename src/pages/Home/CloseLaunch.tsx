import React from "react";
import { Card } from "react-bootstrap";

import LaunchResource from 'api/LaunchResource';

import LaunchStatus from "components/features/LaunchStatusIndicator";

export enum CloseLaunchType {
    Upcoming = "Upcoming",
    Previous = "Previous"
}

interface CloseLaunchProps {
    type: CloseLaunchType;
    launch: LaunchResource;
}

const CloseLaunch = (props: CloseLaunchProps) => {
    return (
        <Card>
            <Card.Header>
                <div className="p-2 pb-3 border-bottom border-light">
                    <h5 className="d-inline">{props.type} launch</h5>
                    {props.launch.status && <div className="float-end"><LaunchStatus status={props.launch.status} /></div>}
                </div>
            </Card.Header>
            <Card.Body>
                <h5>{props.launch.name}</h5>
                <p>{props.launch.mission?.description}</p>
            </Card.Body>
        </Card>
    );
};

export default CloseLaunch;