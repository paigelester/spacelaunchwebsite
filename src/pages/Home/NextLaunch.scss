@import 'App.scss';

.next-launch {
    position: relative;

    &-image-container {
        width: 100%;
        max-height: 90vh;
        overflow: hidden;

        img {
            max-width: 100%;
            height: auto;
        }
    }

    &-details {
        position: absolute;
        width: 100%;
        background-color: $brand-panel-colour;

        .badge {
            color: $black;
            font-weight: normal;
            background-color: $brand-yellow;
        }

        p {
            color: $sub-text-color;
        }

        
    }
}