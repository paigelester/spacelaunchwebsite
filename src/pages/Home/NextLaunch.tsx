import React from "react";
import { Container, Row, Col, Image, Card } from "react-bootstrap";

import LaunchResource from 'api/LaunchResource';

import Countdown from "components/features/Countdown";
import LaunchStatusIndicator from "components/features/LaunchStatusIndicator";

const NextLaunch = (props: LaunchResource) => {
    return (
        <div className='shadow-lg'>
            <Container fluid>
                <Row>
                    <Col md={12} lg={9} className="g-0">
                        <div className="position-relative overflow-hidden">
                            <Image src={props.image} alt='launch' className="img-fluid" />
                            <div className="position-absolute top-0 w-100 h-100 o-75 bg-dark"></div>
                            <div className="position-absolute top-50 start-50 end-0 translate-middle text-center w-100">
                                <p className="text-uppercase text-center mb-1 fs-2">Time to launch</p>
                                <Countdown targetDate={props.window_start} />
                            </div>
                        </div>
                    </Col>
                    <Col className="g-0">
                        <Card className="h-100">
                            <Card.Body>
                                <div className="float-start">
                                    <p className="d-inline fs-6 text-muted border-end border-dark pe-2 me-2">{props.launch_service_provider?.name}</p>
                                    <p className="d-inline fs-6 text-muted">{props.launch_service_provider?.type}</p>
                                </div>
                                {props.status && <div className="float-end"><LaunchStatusIndicator status={props.status} /></div>}
                                <h4 className="pt-5 pb-2">{props.name}</h4>
                                <p>{props.mission?.description}</p>
                            </Card.Body>
                        </Card>
                    </Col>
                </Row>
            </Container>
        </div>
    );
};

export default NextLaunch;