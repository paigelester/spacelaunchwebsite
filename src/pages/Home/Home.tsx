import React from 'react';
import { useResource } from 'rest-hooks';
import { Container, Row, Col } from 'react-bootstrap';

import { UpcomingLaunchResource, PreviousLaunchResource } from 'api/LaunchResource';

import NextLaunch from './NextLaunch';
import CloseLaunch, { CloseLaunchType } from './CloseLaunch';

const HomePage = () => {

    const { results } = useResource(UpcomingLaunchResource.list(), { limit: 2 });
    const previousLaunches = useResource(PreviousLaunchResource.list(), { limit: 1 });

    return (
        <div id='home-page'>
            {results.length > 0 && (
                <div>
                    <NextLaunch
                        name={results[0].name}
                        image={results[0].image}
                        status={results[0].status!.name}
                        launchWindowStart={results[0].window_start}
                        missionDescription={results[0].mission!.description}
                        launchProviderName={results[0].launch_service_provider!.name}
                        launchProviderType={results[0].launch_service_provider!.type} />
                    <Container fluid={true} className="">
                        <Row>
                            <Col sm={12} md={6} className="p-0">
                                <CloseLaunch
                                    name={previousLaunches.results[0].name}
                                    type={CloseLaunchType.Previous}
                                    status={previousLaunches.results[0].status!.name}
                                    description={previousLaunches.results[0].mission!.description} />
                            </Col>
                            <Col sm={12} md={6} className="p-0">
                                <CloseLaunch
                                    name={results[1].name}
                                    type={CloseLaunchType.Upcoming}
                                    status={results[1].status!.name}
                                    description={results[1].mission!.description} />
                            </Col>
                        </Row>
                    </Container>
                </div>
            )}
            <div>
                News (Upcoming events, expedictions, landings, launches)
            </div>
            <div>
                Agencies
            </div>
        </div>
    );
};

export default HomePage;