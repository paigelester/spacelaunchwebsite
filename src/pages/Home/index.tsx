import React from 'react';
import { useResource } from 'rest-hooks';
import { Container, Row, Col } from 'react-bootstrap';

import LaunchResource from 'api/LaunchResource';

import NextLaunch from './NextLaunch';
import CloseLaunch, { CloseLaunchType } from './CloseLaunch';

const HomePage = () => {

    const upcomingLaunches = useResource(LaunchResource.upcomingLaunches(), { limit: 2 });
    const previousLaunches = useResource(LaunchResource.upcomingLaunches(), { limit: 1 });
    
    return (
        <div id='home-page'>
            {upcomingLaunches.count > 0 && (
                <div data-testid='launches'>
                    <NextLaunch {...upcomingLaunches.results[0]} />
                    <Container fluid={true} className="">
                        <Row>
                            <Col sm={12} md={6} className="p-0">
                                <CloseLaunch
                                    launch={previousLaunches.results[0]}
                                    type={CloseLaunchType.Previous}/>
                            </Col>
                            <Col sm={12} md={6} className="p-0">
                                <CloseLaunch
                                    launch={upcomingLaunches.results[0]}
                                    type={CloseLaunchType.Upcoming} />
                            </Col>
                        </Row>
                    </Container>
                </div>
            )}
            <div>
                News (Upcoming events, expedictions, landings, launches)
            </div>
            <div>
                Agencies
            </div>
        </div>
    );
};

export default HomePage;