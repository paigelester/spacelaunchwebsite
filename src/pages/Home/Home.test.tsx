import React, { Suspense } from 'react';
import { CacheProvider } from 'rest-hooks';
import { render, waitFor } from '@testing-library/react';
import { makeServer } from 'mirage/server';
import Home from './index';

let server: any;
beforeEach(() => {
    server = makeServer("test");
    Date.now = jest.fn(() => 1618171860892);
});
afterEach(() => { server.shutdown() });

it("renders as expected", async () => {
    // given
    server.createList("launch", 2, "upcoming");
    server.create("launch");

    // when
    const { container, getByTestId } = render(
        <CacheProvider>
            <Suspense fallback={<div>Loading...</div>}>
                <Home />
            </Suspense>
        </CacheProvider>
    );

    await waitFor(() => getByTestId('launches'));

    // then
    expect(container).toMatchSnapshot();
});

it("suspends then resolves", async () => {
    // given
    const loadingText: string = "Loading...";
    server.createList("launch", 2, "upcoming");
    server.create("launch");

    // when
    const { queryByText, getByTestId } = render(
        <CacheProvider>
            <Suspense fallback={<div>Loading...</div>}>
                <Home />
            </Suspense>
        </CacheProvider>
    );

    // then
    expect(queryByText(loadingText)).toBeDefined();
    expect(await waitFor(() => getByTestId('launches')));
});
