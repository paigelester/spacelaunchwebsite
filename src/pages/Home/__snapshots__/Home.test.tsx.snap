// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    id="home-page"
  >
    <div
      data-testid="launches"
    >
      <div
        class="shadow-lg"
      >
        <div
          class="container-fluid"
        >
          <div
            class="row"
          >
            <div
              class="g-0 col-lg-9 col-md-12"
            >
              <div
                class="position-relative overflow-hidden"
              >
                <img
                  alt="launch"
                  class="img-fluid"
                  src="https://spacelaunchnow-prod-east.nyc3.digitaloceanspaces.com/media/launch_images/falcon2520925_image_20210308112429.png"
                />
                <div
                  class="position-absolute top-0 w-100 h-100 o-75 bg-dark"
                />
                <div
                  class="position-absolute top-50 start-50 end-0 translate-middle text-center w-100"
                >
                  <p
                    class="text-uppercase text-center mb-1 fs-2"
                  >
                    Time to launch
                  </p>
                  <div
                    class="countdown"
                  >
                    <div
                      class="text-center me-2 d-inline-block"
                    >
                      <p
                        class="countdown-days-value fs-1 m-0"
                      >
                        18
                      </p>
                      <p
                        class="countdown-days text-muted text-uppercase m-0"
                      >
                        days
                      </p>
                    </div>
                    <div
                      class="text-center me-2 d-inline-block"
                    >
                      <p
                        class="countdown-hours-value fs-1 m-0"
                      >
                        03
                      </p>
                      <p
                        class="countdown-hours text-muted text-uppercase m-0"
                      >
                        hours
                      </p>
                    </div>
                    <div
                      class="text-center me-2 d-inline-block"
                    >
                      <p
                        class="countdown-minutes-value fs-1 m-0"
                      >
                        48
                      </p>
                      <p
                        class="countdown-minutes text-muted text-uppercase m-0"
                      >
                        minutes
                      </p>
                    </div>
                    <div
                      class="text-center d-inline-block"
                    >
                      <p
                        class="countdown-seconds-value fs-1 m-0"
                      >
                        59
                      </p>
                      <p
                        class="countdown-seconds text-muted text-uppercase m-0"
                      >
                        seconds
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="g-0 col"
            >
              <div
                class="h-100 card"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="float-start"
                  >
                    <p
                      class="d-inline fs-6 text-muted border-end border-dark pe-2 me-2"
                    >
                      Launch provider 1
                    </p>
                    <p
                      class="d-inline fs-6 text-muted"
                    >
                      Government
                    </p>
                  </div>
                  <div
                    class="float-end"
                  >
                    <span
                      class="badge bg-warning"
                    >
                      TBD
                    </span>
                  </div>
                  <h4
                    class="pt-5 pb-2"
                  >
                    Falcon 9 Block 5 | Starlink 21
                  </h4>
                  <p>
                    First cargo delivery mission to the space station.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="m-4"
      >
        <h2
          class="mb-4"
        >
          Upcoming launches
        </h2>
        <div
          class="react-multi-carousel-list  "
        >
          <ul
            class="react-multi-carousel-track "
            style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
